sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala

```diff
@@ -419,14 +441,44 @@ class StreamExecution(
             SQLConf.SHUFFLE_PARTITIONS.key, shufflePartitionsToUse.toString)
         }
 
-        logDebug(s"Found possibly unprocessed offsets $availableOffsets " +
-          s"at batch timestamp ${offsetSeqMetadata.batchTimestampMs}")
+        offsetLog.get(latestBatchId - 1).foreach { lastOffsets =>
+          committedOffsets = lastOffsets.toStreamProgress(sources)
+        }
 
-        offsetLog.get(batchId - 1).foreach {
-          case lastOffsets =>
-            committedOffsets = lastOffsets.toStreamProgress(sources)
-            logDebug(s"Resuming with committed offsets: $committedOffsets")
+        /* identify the current batch id: if commit log indicates we successfully processed the
+         * latest batch id in the offset log, then we can safely move to the next batch
+         * i.e., committedBatchId + 1
+         */
+        batchCommitLog.getLatest() match {
+          case Some((completionBatchId, _))
```

Comment:
I would write the conditions as 
```
case Some((completionBatchId, _)) => 
  if (completionBatchId == latestBatchId) { ... }
  else if (completionBatchId < latestBatchId - 1) { logWarning(...) }
case None => 
```
It is less intuitive to look at the current cases and see whether they are exhaustive or not.